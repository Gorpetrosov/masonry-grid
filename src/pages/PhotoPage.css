/* PhotoPage.css */
.cart .container {
    max-width: 1400px;
    margin: 0 auto;
    padding: 0 1rem;
}

.cart .center {
    text-align: center;
}

.cart-wrapper {
    background: #ffffff;
    border-radius: 12px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    padding: 1.5rem;
    margin: 1rem 0;
}

.cart-content {
    display: flex;
    flex-direction: column;
    gap: 1.5rem;
}

.cart-image {
    position: relative;
    overflow: hidden;
    aspect-ratio: 16/9;
    margin-bottom: 20px;
    break-inside: avoid;
    border-radius: 15px;
    transition: transform 0.3s ease;
    box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.3);
    padding: 0;
}

.cart-image img {
    width: 100%;
    height: 100%;
    object-fit: cover;
    display: block;
    transition: transform 0.3s ease;
}
.cart-image:hover img {
    transform: scale(1.1);
}

.cart-detail {
    padding: 0.5rem;
}

.text-medium {
    font-size: 1.25rem;
    color: #2d3748;
    margin-bottom: 1rem;
    text-align: center;
}

.text-base {
    font-size: 1.1rem;
    color: #4a5568;
    margin-bottom: 0.75rem;
}

.cart-detail p {
    font-size: 0.95rem;
    color: #718096;
    margin: 0.5rem 0;
    line-height: 1.5;
}



.cart-checkout {
    border-top: 1px solid #e2e8f0;
    padding-top: 1rem;
    padding-bottom: 3rem;
    margin-top: 1rem;
    text-align: center;
}

/* ============ Responsive Breakpoints ============ */
@media (min-width: 480px) {
    .cart-image {
        aspect-ratio: 4/3;
    }

    .text-medium {
        font-size: 1.35rem;
    }
}

@media (min-width: 768px) {

    .cart-content {
        flex-direction: row;
        gap: 2.5rem;
    }

    .cart-image {
        flex: 1 1 60%;
        aspect-ratio: 16/9;
    }

    .cart-detail {
        flex: 1 1 40%;
        padding: 1rem;
    }

    .text-medium {
        font-size: 1.5rem;
        text-align: left;
    }

    .text-base {
        font-size: 1.25rem;
    }

    .cart-detail p {
        font-size: 1rem;
    }
}

@media (min-width: 1024px) {
    .container {
        padding: 0 2rem;
    }

    .cart-wrapper {
        padding: 2.5rem;
    }

    .cart-content {
        gap: 3rem;
    }

    .text-medium {
        font-size: 1.75rem;
    }
}

@media (min-width: 1280px) {
    .cart-image {
        flex: 1 1 70%;
    }

    .cart-detail {
        flex: 1 1 30%;
    }
}

/* Touch devices */
@media (hover: none) {
    .cart-image:hover img {
        transform: none;
    }

    .custom-link:hover {
        transform: none;
        background-color: transparent;
    }
}

/* High contrast mode */
@media (prefers-contrast: more) {
    .cart-wrapper {
        border: 2px solid #000;
    }

    .custom-link {
        text-decoration: underline;
    }
}