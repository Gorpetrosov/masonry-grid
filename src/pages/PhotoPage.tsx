import Layout from "../components/layout/Layout.tsx";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../store/store.ts";
import { fetchPhoto } from "../store/photoSlice.ts";
import { Link, useParams } from "react-router-dom";
import { useEffect, useMemo } from "react";
import "./PhotoPage.css";
import NotFoundPage from "./404/NotFoundPage.tsx";
import { Photo } from "../interfaces/photo.ts";

const PhotoPage = () => {
    const dispatch = useDispatch<AppDispatch>();
    const { entities, entity } = useSelector((state: RootState) => state.photos);
    const { id } = useParams<{ id: string }>();
    const photoId = id ? Number(id) : 0;

    const photo = useMemo<Photo | null>(
        () => entities.find((p) => p.id === photoId) || entity,
        [entities, entity, photoId]
    );

    useEffect(() => {
        if (!photo && photoId) {
            dispatch(fetchPhoto(photoId));
        }
    }, [dispatch, photo, photoId]);

    return (
        <Layout>
            {photo ? (
                <main className="cart main">
                    <section className="cart container" id="cart">
                        <h3 className="text-medium font-medium center">{photo.alt}</h3>
                        <div className="cart-wrapper">
                            <article className="cart-content">
                                <div className="cart-image">
                                    <img src={photo.src.original} alt={photo.alt} />
                                </div>
                                <div className="cart-detail">
                                    <h3 className="text-base font-semi">Photographer {photo.photographer}</h3>
                                    <Link className="custom-link" target="_blank" to={photo.photographer_url}>
                                        Link to {photo.photographer}'s page
                                    </Link>
                                    <p>Original Width: {photo.width} px</p>
                                    <p>Original Height: {photo.height} px</p>
                                </div>
                            </article>
                        </div>
                        <div className="cart-checkout">
                            <Link className="custom-link" to="/">Main Page</Link>
                        </div>
                    </section>
                </main>
            ) : (
                <NotFoundPage text={`There is no image with id ${id}`} />
            )}
        </Layout>
    );
};

export default PhotoPage;
