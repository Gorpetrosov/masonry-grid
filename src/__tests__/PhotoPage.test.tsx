import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import {BrowserRouter, useParams} from 'react-router-dom'
import PhotoPage from '../pages/PhotoPage.tsx'
import photosReducer, { fetchPhoto } from '../store/photoSlice.ts' // Added fetchPhoto import
import type { Photo } from '../interfaces/photo.ts'
import { vi } from 'vitest'
import {ReactNode} from "react";

vi.mock('../components/layout/Layout.tsx', () => ({
    default: ({ children }: { children: ReactNode }) => (
        <div data-testid="layout-mock">{children}</div>
    )
}))

vi.mock('./404/NotFoundPage.tsx', () => ({
    default: ({ text }: { text: string }) => (
        <div data-testid="not-found-mock">{text}</div>
    )
}))

vi.mock('react-router-dom', async (importOriginal) => ({
    ...(await importOriginal<typeof import('react-router-dom')>()),
    useParams: vi.fn(() => ({ id: '123' })),
}))

vi.mock('../store/photoSlice', async (importOriginal) => {
    const actual = await importOriginal<typeof import('../store/photoSlice.ts')>();
    return {
        ...actual,
        fetchPhoto: vi.fn((id: number) => ({
            type: 'MOCK_FETCH_PHOTO',
            payload: id,
        })),
    };
});

const mockPhoto: Photo = {
    id: 123,
    alt: 'Brown Rocks During Golden Hour',
    photographer: 'John Doe',
    src: {
        tiny: 'test.jpg',
        original: 'test-original.jpg',
        small: '',
        medium: '',
        large: '',
        portrait: '',
        landscape: '',
        large2x: ''
    },
    photographer_url: 'https://example.com',
    width: 1920,
    height: 1080,
    url: '',
    liked: false,
    photographer_id: 1,
    avg_color: 'jjj'
}

const createTestStore = (preloadedState: any) => {
    const store = configureStore({
        reducer: {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-expect-error
            photos: photosReducer
        },
        preloadedState
    })
    const dispatchSpy = vi.spyOn(store, 'dispatch')
    return { store, dispatchSpy }
}

const renderComponent = (store: any) => {
    return render(
        <Provider store={store}>
            <BrowserRouter>
                <PhotoPage />
            </BrowserRouter>
        </Provider>
    )
}

describe('PhotoPage', () => {
    beforeEach(() => {
        vi.clearAllMocks()
        vi.mocked(useParams).mockReturnValue({ id: '123' })
    })

    it('should match snapshot when photo exists', () => {
        const { store } = createTestStore({
            photos: {
                entities: [mockPhoto],
                entity: null,
                loading: false,
                error: null
            }
        })

        const { container } = renderComponent(store)
        expect(container).toMatchSnapshot()
    })

    it('should match snapshot when photo not found', () => {
        const { store } = createTestStore({
            photos: {
                entities: [],
                entity: null,
                loading: false,
                error: null
            }
        })

        const { container } = renderComponent(store)
        expect(container).toMatchSnapshot()
    })

    it('should dispatch fetchPhoto when photo not in cache', () => {
        const { store, dispatchSpy } = createTestStore({
            photos: {
                entities: [],
                entity: null,
                loading: false,
                error: null
            }
        })

        renderComponent(store)
        expect(dispatchSpy).toHaveBeenCalledWith(fetchPhoto(123))
    })
})