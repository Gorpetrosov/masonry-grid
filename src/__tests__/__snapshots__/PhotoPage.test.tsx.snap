// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PhotoPage > should match snapshot when photo exists 1`] = `
<div>
  <div
    data-testid="layout-mock"
  >
    <main
      class="cart main"
    >
      <section
        class="cart container"
        id="cart"
      >
        <h3
          class="text-medium font-medium center"
        >
          Brown Rocks During Golden Hour
        </h3>
        <div
          class="cart-wrapper"
        >
          <article
            class="cart-content"
          >
            <div
              class="cart-image"
            >
              <img
                alt="Brown Rocks During Golden Hour"
                src="test-original.jpg"
              />
            </div>
            <div
              class="cart-detail"
            >
              <h3
                class="text-base font-semi"
              >
                Photographer 
                John Doe
              </h3>
              <a
                class="custom-link"
                href="https://example.com"
                target="_blank"
              >
                Link to 
                John Doe
                's page
              </a>
              <p>
                Original Width: 
                1920
                 px
              </p>
              <p>
                Original Height: 
                1080
                 px
              </p>
            </div>
          </article>
        </div>
        <div
          class="cart-checkout"
        >
          <a
            class="custom-link"
            data-discover="true"
            href="/"
          >
            Main Page
          </a>
        </div>
      </section>
    </main>
  </div>
</div>
`;

exports[`PhotoPage > should match snapshot when photo not found 1`] = `
<div>
  <div
    data-testid="layout-mock"
  >
    <div
      class="container-404"
    >
      <h1>
        404 Not Found
      </h1>
      <p
        class="zoom-area"
      >
        There is no image with id 123
      </p>
      <p
        class="zoom-area"
      >
         Go back to the homepage if you dare!
      </p>
      <section
        class="error-container"
      >
        <span>
          4
        </span>
        <span>
          <span
            class="screen-reader-text"
          >
            0
          </span>
        </span>
        <span>
          4
        </span>
      </section>
      <div
        class="link-container"
      >
        <a
          class="custom-link"
          data-discover="true"
          href="/"
        >
          GO TO MAIN PAGE
        </a>
      </div>
    </div>
  </div>
</div>
`;
